File: BuildTree-inorder-postorder.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

#Accepted by leetcode
#Time complexity - O(n) searching root value in inorder
#Space complexity - O(n) asymtotically because we are using extra arrays inleft, inright, postleft, postright

class Solution(object):
    def buildTree(self, inorder, postorder):
        """
        :type inorder: List[int]
        :type postorder: List[int]
        :rtype: TreeNode
        """
        #Edge case
        if len(postorder) == 0:
            return None        
        root = TreeNode(postorder[-1])
        rootindx = -1
        for i,val in enumerate(inorder):
            if val == root.val:
                rootindx = i
        inleft = inorder[:rootindx]
        inright = inorder[rootindx+1:]
        postleft = postorder[:rootindx]
        postright = postorder[rootindx:-1]
        root.left = self.buildTree(inleft, postleft)
        root.right = self.buildTree(inright, postright)
        return root            

File: buildTreeInOrderPostOrder.java

//https://leetcode.com/problems/construct-binary-tree-from-inorder-and-postorder-traversal/
/*
Time: O(n)
Space: O(n)
Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : None
*/
class Solution{

    HashMap<Integer, Integer> map = new HashMap();
    int idx;

    public TreeNode buildTree(int[] inorder, int[] postorder) {
        idx = postorder.length - 1;
        int iStart = 0;
        int iEnd = inorder.length - 1;

        for (int i = 0; i < inorder.length; i++)
            map.put(inorder[i], i); // memoization

        return helper(inorder, postorder, iStart, iEnd);

    }

    public TreeNode helper(int[] inorder, int[] postorder, int iStart, int iEnd) {
        if (iStart > iEnd) // if there is no elements to construct subtrees
            return null;

        TreeNode root = new TreeNode(postorder[idx]);

        int iRoot = map.get(postorder[idx]); // pick up post_idx element as a root
        // recursion 
        idx--;
          // build right subtree
        root.right = helper(inorder, postorder, iRoot + 1, iEnd);
         // build left subtree
        root.left = helper(inorder, postorder, iStart, iRoot - 1);

        return root;

    }

}